nexus/db-queries/src/db/datastore/ip_pool.rs

```diff
@@ -46,29 +56,108 @@ impl DataStore {
+                ip_pool::name,
                 &pagparams.map_name(|n| Name::ref_cast(n)),
             ),
         }
-        // != excludes nulls so we explicitly include them
-        .filter(dsl::silo_id.ne(*INTERNAL_SILO_ID).or(dsl::silo_id.is_null()))
-        .filter(dsl::time_deleted.is_null())
+        .left_outer_join(ip_pool_resource::table)
+        .filter(
+            ip_pool_resource::resource_id
+                .ne(*INTERNAL_SILO_ID)
+                // resource_id is not nullable -- null here means the
+                // pool has no entry in the join table
+                .or(ip_pool_resource::resource_id.is_null()),
+        )
+        .filter(ip_pool::time_deleted.is_null())
+        .select(IpPool::as_select())
+        .get_results_async(&*self.pool_connection_authorized(opctx).await?)
+        .await
+        .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))
+    }
+
+    /// List IP pools linked to the current silo
+    pub async fn silo_ip_pools_list(
+        &self,
+        opctx: &OpContext,
+        pagparams: &PaginatedBy<'_>,
+    ) -> ListResultVec<db::model::IpPool> {
+        use db::schema::ip_pool;
+        use db::schema::ip_pool_resource;
+
+        // From the developer user's point of view, we treat IP pools linked to
+        // their silo as silo resources, so they can list them if they can list
+        // silo children
+        let authz_silo =
+            opctx.authn.silo_required().internal_context("listing IP pools")?;
+        opctx.authorize(authz::Action::ListChildren, &authz_silo).await?;
+
+        let silo_id = authz_silo.id();
+
+        match pagparams {
+            PaginatedBy::Id(pagparams) => {
+                paginated(ip_pool::table, ip_pool::id, pagparams)
+            }
+            PaginatedBy::Name(pagparams) => paginated(
+                ip_pool::table,
+                ip_pool::name,
+                &pagparams.map_name(|n| Name::ref_cast(n)),
+            ),
+        }
+        .inner_join(ip_pool_resource::table)
+        .filter(
+            ip_pool_resource::resource_type
+                .eq(IpPoolResourceType::Silo)
+                .and(ip_pool_resource::resource_id.eq(silo_id)),
+        )
+        .filter(ip_pool::time_deleted.is_null())
         .select(db::model::IpPool::as_select())
         .get_results_async(&*self.pool_connection_authorized(opctx).await?)
         .await
         .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))
     }
 
+    /// Look up whether the given pool is available to users in the current
+    /// silo, i.e., whether there is an entry in the association table linking
+    /// the pool with that silo
+    pub async fn ip_pool_fetch_link(
+        &self,
+        opctx: &OpContext,
+        ip_pool_id: Uuid,
+    ) -> LookupResult<IpPoolResource> {
+        use db::schema::ip_pool;
+        use db::schema::ip_pool_resource;
+
+        let authz_silo = opctx.authn.silo_required()?;
```

Comment:
why no `internal_context("...")` call here? I think that gets spit out in the logs, can be useful.